docs/Developers/NextGen_Core.rst

```diff
@@ -0,0 +1,303 @@
+their corresponding results (in the form of a list) as *Key-Value* pairs.
+From the previous example if we try to access the BEAR_DEPS of the ``BarBear``
+we will get the result ``{<class 'coalib.core.Bear.FooBear'>}``.
+
+In the `__init__()` method of the class ``Bear.py`` the dependencies specified
+in the ``BEAR_DEPS`` are copied to every instance of a Bear run using
+``self.BEAR_DEPS = set(self.BEAR_DEPS)`` which makes runtime modifications
+possible.
+
+Override bears
+--------------
+
+A NextGen bear has to have the following functions to perform analysis:
+
+- ``analyze``: This method contains the code that performs the actual code
+  analysis routine that that bear is used for.
+
+- ``generate_tasks``: This method is a part of the parent ``Bear.py`` class
+  and returns tuples containing the positional arguments as a tuple and the
+  keyword arguments in the form of a dict. These are actually the task objects
+  that are scheduled and executed by the core. An absence of this method raises
+  ``NotImplementedError``. For example if we instantiate an object of
+  ``FileBear`` and call the ``generate_tasks`` method we would get a tuple of a
+  tuple of the filename and the file in the file_dict and a dictionary of kwargs
+  to the ``analyze`` method.
+
+  ::
+
+      from coalib.core.FileBear import FileBear
+
+      section = Section('test-section')
+      file_dict = {'fileA.py': []}
+
+      FileBear(section, file_dict).generate_tasks()
+
+In the same way ``generate_tasks`` can be used in bears inheriting from
+``ProjectBear`` or ``DependencyBear``.
+
+Let's write our own bears with custom ``generate_tasks`` methods.
+
+::
+
+    # This bear provides some code analysis
+    class SomeBear(Bear):
+
+        def analyze(self, bear, section_name, file_dict):
+            yield "Some analysis"
+
+    # This bear depends on the above bear and performs some
+    # more analysis after receving its results
+    class SomeOtherBear(Bear):
+        BEAR_DEPS = {SomeBear}
+
+        def analyze(self, *args):
+            return "More analysis"
+
+        def generate_tasks(self):
+            result_length = sum(len(results)
+                                for results in self.dependency_results.values())
+            return (((i,), {}) for i in range(result_length))
+
+Running the Core
+----------------
+
+The `run(bears, result_callback, cache=None, executor=None)` method in
+``Core.py`` is used to set up a coala session and run it.
+- ``bears``: contains all the instances of bears to run.
+- ``result_callback``: a function that is called only when the results from
+the bears are available.
+- ``cache``: a cache used by bears to speed up runs.
+- ``executor``: a custom executor used to run the bears. By default
+``ProcessPoolExecutor`` is used.
+
+The new core comes with the following new additions:
+
+- **Dependency Resolution:** The ``initialize_dependencies`` method in
```

Comment:
There are already 4 paragraphs above that talk in details about the new Dependency Resolution scheme of the new core. IMHO the contents of this paragraph should be absorbed in them and this one should be removed.

Reply:
Read the original issue comment please https://github.com/coala/coala/issues/4584
It mentions about 3 to 4 sections about dependency resolution so it isn't redundant.